File: hftbacktest/local.py
from numba import typeof, int64, float64
from numba.experimental import jitclass
from numba.typed.typeddict import Dict

from .reader import COL_EVENT, COL_LOCAL_TIMESTAMP, COL_SIDE, COL_PRICE, COL_QTY, DEPTH_CLEAR_EVENT, DEPTH_EVENT, \
    DEPTH_SNAPSHOT_EVENT
from .marketdepth import MarketDepth
from .order import BUY, SELL, NEW, CANCELED, FILLED, EXPIRED, NONE, Order, order_ty, OrderBus, order_ladder_ty


class Local_:
    def __init__(
            self,
            data_reader,
            orders_to_exch,
            orders_from_exch,
            depth,
            state,
            order_latency
    ):
        self.data_reader = data_reader
        self.data = data_reader.next()
        self.next_data = data_reader.next()
        self.row_num = -1

        self.orders = Dict.empty(int64, order_ty)

        self.orders_to = orders_to_exch
        self.orders_from = orders_from_exch

        self.depth = depth
        self.state = state
        self.order_latency = order_latency

    def next_timestamp(self):
        next_data_timestamp = self.__next_data_timestamp()
        next_recv_order_timestamp = self.orders_from.frontmost_timestamp

        if (0 < next_recv_order_timestamp < next_data_timestamp) \
                or (next_data_timestamp <= 0 < next_recv_order_timestamp):
            return next_recv_order_timestamp
        else:
            return next_data_timestamp

    def __next_data_timestamp(self):
        if self.row_num + 1 < len(self.data):
            return self.data[self.row_num + 1, COL_LOCAL_TIMESTAMP]
        else:
            if len(self.next_data) == 0:
                return -2
            return self.next_data[0, COL_LOCAL_TIMESTAMP]

    def __process_recv_order(self, wait_resp):
        next_frontmost_timestamp = 0
        i = 0
        while i < self.orders_from.__len__():
            order, recv_timestamp = self.orders_from[i]
            if self.orders_from.frontmost_timestamp == recv_timestamp:
                self.orders_from.__delitem__(i)

                # Apply the received order response to the local orders.
                self.orders[order.order_id] = order
                if order.status == FILLED:
                    self.state.apply_fill(order)
            else:
                i += 1
                # Find the next frontmost timestamp
                if next_frontmost_timestamp <= 0:
                    next_frontmost_timestamp = recv_timestamp
                else:
                    next_frontmost_timestamp = min(next_frontmost_timestamp, recv_timestamp)
        self.orders_from.frontmost_timestamp = next_frontmost_timestamp
        return 0

    def __process_data(self, wait_resp):
        # Move to the next row.
        self.row_num += 1
        if self.row_num == len(self.data):
            self.data_reader.release(self.data)
            self.data = self.next_data
            self.next_data = self.data_reader.next()
            self.row_num = 0

        row = self.data[self.row_num]

        # Process a depth event
        if row[COL_EVENT] == DEPTH_CLEAR_EVENT:
            self.depth.clear_depth(row[COL_SIDE], row[COL_PRICE])
        elif row[COL_EVENT] == DEPTH_EVENT or row[COL_EVENT] == DEPTH_SNAPSHOT_EVENT:
            if row[COL_SIDE] == BUY:
                self.depth.update_bid_depth(
                    row[COL_PRICE],
                    row[COL_QTY],
                    row[COL_LOCAL_TIMESTAMP]
                )
            else:
                self.depth.update_ask_depth(
                    row[COL_PRICE],
                    row[COL_QTY],
                    row[COL_LOCAL_TIMESTAMP]
                )

        return 0

    def process(self, wait_resp):
        next_data_timestamp = self.__next_data_timestamp()
        next_recv_order_timestamp = self.orders_from.frontmost_timestamp

        if (0 < next_recv_order_timestamp < next_data_timestamp) \
                or (next_data_timestamp <= 0 < next_recv_order_timestamp):
            return self.__process_recv_order(wait_resp)
        else:
            return self.__process_data(wait_resp)

    def submit_buy_order(self, order_id, price, qty, time_in_force, current_timestamp):
        price_tick = round(price / self.depth.tick_size)
        order = Order(order_id, price_tick, self.depth.tick_size, qty, BUY, time_in_force)
        order.req = NEW
        exch_recv_timestamp = current_timestamp + self.order_latency.entry(order, self)

        self.orders[order.order_id] = order
        self.orders_to.append(order.copy(), exch_recv_timestamp)

    def submit_sell_order(self, order_id, price, qty, time_in_force, current_timestamp):
        price_tick = round(price / self.depth.tick_size)
        order = Order(order_id, price_tick, self.depth.tick_size, qty, SELL, time_in_force)
        order.req = NEW
        exch_recv_timestamp = current_timestamp + self.order_latency.entry(order, self)

        self.orders[order.order_id] = order
        self.orders_to.append(order.copy(), exch_recv_timestamp)

    def cancel(self, order_id, current_timestamp):
        order = self.orders.get(order_id)

        if order is None:
            raise KeyError('the given order_id does not exist.')
        if order.req != NONE:
            raise ValueError('the given order cannot be cancelled because there is a ongoing request.')

        order.req = CANCELED
        exch_recv_timestamp = current_timestamp + self.order_latency.entry(order, self)

        self.orders_to.append(order.copy(), exch_recv_timestamp)

    def clear_inactive_orders(self):
        for order in list(self.orders.values()):
            if order.status == EXPIRED \
                    or order.status == FILLED \
                    or order.status == CANCELED:
                del self.orders[order.order_id]


def Local(
        data_reader,
        orders_to_exch,
        orders_from_exch,
        depth,
        state,
        order_latency
):
    jitted = jitclass(spec=[
        ('data_reader', typeof(data_reader)),
        ('data', float64[:, :]),
        ('next_data', float64[:, :]),
        ('row_num', int64),

        ('orders', order_ladder_ty),

        ('orders_to', OrderBus.class_type.instance_type),
        ('orders_from', OrderBus.class_type.instance_type),

        ('depth', MarketDepth.class_type.instance_type),
        ('state', typeof(state)),
        ('order_latency', typeof(order_latency)),
    ])(Local_)
    return jitted(
        data_reader,
        orders_to_exch,
        orders_from_exch,
        depth,
        state,
        order_latency
    )

File: rust/src/proc/local.rs
use std::collections::hash_map::Entry;
use std::collections::HashMap;
use std::mem;
use crate::assettype::AssetType;
use crate::depth::MarketDepth;
use crate::Error;
use crate::models::LatencyModel;
use crate::order::{Order, OrderBus, OrdType, Side, Status, TimeInForce};
use crate::proc::proc::{LocalProcessor, Processor};
use crate::reader::{
    BUY,
    SELL,
    LOCAL_ASK_DEPTH_CLEAR_EVENT,
    LOCAL_ASK_DEPTH_EVENT,
    LOCAL_ASK_DEPTH_SNAPSHOT_EVENT,
    LOCAL_BID_DEPTH_CLEAR_EVENT,
    LOCAL_BID_DEPTH_EVENT,
    LOCAL_BID_DEPTH_SNAPSHOT_EVENT,
    LOCAL_EVENT,
    LOCAL_TRADE_EVENT,
    Reader,
    Data,
    Row
};
use crate::state::State;

pub struct Local<A, Q, L>
    where
        A: AssetType,
        Q: Clone,
        L: LatencyModel
{
    pub reader: Reader<Row>,
    pub data: Data<Row>,
    row_num: usize,

    pub orders: HashMap<i64, Order<Q>>,

    pub orders_to: OrderBus<Q>,
    pub orders_from: OrderBus<Q>,

    pub depth: MarketDepth,
    pub state: State<A>,
    order_latency: L,

    trades: Vec<Row>
}

impl<A, Q, L> Local<A, Q, L>
    where
        A: AssetType,
        Q: Clone + Default,
        L: LatencyModel
{
    pub fn new(
        reader: Reader<Row>,
        depth: MarketDepth,
        state: State<A>,
        order_latency: L,
        trade_len: usize,
        orders_to: OrderBus<Q>,
        orders_from: OrderBus<Q>
    ) -> Self {
        Self {
            reader,
            data: Data::empty(),
            row_num: 0,
            orders: Default::default(),
            orders_to,
            orders_from,
            depth,
            state,
            order_latency,
            trades: Vec::with_capacity(trade_len),
        }
    }

    fn process_recv_order_(
        &mut self,
        order: Order<Q>,
        _recv_timestamp: i64,
        _wait_resp: i64,
        next_timestamp: i64
    ) -> Result<i64, Error> {
        if order.status == Status::Filled {
            self.state.apply_fill(&order);
        }
        // Applies the received order response to the local orders.
        match self.orders.entry(order.order_id) {
            Entry::Occupied(mut entry) => {
                *entry.get_mut() = order;
            }
            Entry::Vacant(entry) => {
                entry.insert(order);
            }
        }

        // Bypass next_timestamp
        Ok(next_timestamp)
    }

    pub fn clear_last_trades(&mut self) {
        self.trades.clear();
    }
}

impl<A, Q, L> LocalProcessor<A, Q> for Local<A, Q, L>
    where
        A: AssetType,
        Q: Clone + Default,
        L: LatencyModel
{
    fn submit_order(
        &mut self,
        order_id: i64,
        side: Side,
        price: f32,
        qty: f32,
        order_type: OrdType,
        time_in_force: TimeInForce,
        current_timestamp: i64
    ) -> Result<(), Error> {
        if self.orders.contains_key(&order_id) {
            return Err(Error::OrderAlreadyExist);
        }

        let price_tick = (price / self.depth.tick_size).round() as i32;
        let mut order = Order::new(
            order_id,
            price_tick,
            self.depth.tick_size,
            qty,
            side,
            order_type,
            time_in_force
        );
        order.req = Status::New;
        let exch_recv_timestamp = current_timestamp + self.order_latency.entry(current_timestamp, &order);

        self.orders_to.append(order.clone(), exch_recv_timestamp);
        self.orders.insert(order.order_id, order);
        Ok(())
    }

    fn cancel(&mut self, order_id: i64, current_timestamp: i64) -> Result<(), Error> {
        let order = self.orders.get_mut(&order_id).ok_or(Error::OrderNotFound)?;

        if order.req != Status::None {
            return Err(Error::OrderRequestInProcess);
        }

        order.req = Status::Canceled;
        let exch_recv_timestamp = current_timestamp + self.order_latency.entry(current_timestamp, order);

        self.orders_to.append(order.clone(), exch_recv_timestamp);
        Ok(())
    }

    fn clear_inactive_orders(&mut self) {
        self.orders.retain(
            |_, order| order.status != Status::Expired
                && order.status != Status::Filled
                && order.status != Status::Canceled
        )
    }

    fn state(&self) -> &State<A> {
        &self.state
    }

    fn depth(&self) -> &MarketDepth {
        &self.depth
    }

    fn orders(&self) -> &HashMap<i64, Order<Q>> {
        &self.orders
    }
}

impl<A, Q, L> Processor for Local<A, Q, L>
    where
        A: AssetType,
        Q: Clone + Default,
        L: LatencyModel
{
    fn initialize_data(&mut self) -> Result<i64, Error> {
        self.data = self.reader.next()?;
        for rn in 0..self.data.len() {
            if self.data[rn].ev & LOCAL_EVENT == LOCAL_EVENT {
                self.row_num = rn;
                return Ok(self.data[rn].local_ts);
            }
        }
        Err(Error::EndOfData)
    }

    fn process_data(&mut self) -> Result<(i64, i64), Error> {
        let row = &self.data[self.row_num];
        // Processes a depth event
        if row.ev & LOCAL_BID_DEPTH_CLEAR_EVENT == LOCAL_BID_DEPTH_CLEAR_EVENT {
            self.depth.clear_depth(BUY, row.px);
        } else if row.ev & LOCAL_ASK_DEPTH_CLEAR_EVENT == LOCAL_ASK_DEPTH_CLEAR_EVENT {
            self.depth.clear_depth(SELL, row.px);
        } else if row.ev & LOCAL_BID_DEPTH_EVENT == LOCAL_BID_DEPTH_EVENT
            || row.ev & LOCAL_BID_DEPTH_SNAPSHOT_EVENT == LOCAL_BID_DEPTH_SNAPSHOT_EVENT {
            self.depth.update_bid_depth(
                row.px,
                row.qty,
                row.local_ts
            );
        } else if row.ev & LOCAL_ASK_DEPTH_EVENT == LOCAL_ASK_DEPTH_EVENT
            || row.ev & LOCAL_ASK_DEPTH_SNAPSHOT_EVENT == LOCAL_ASK_DEPTH_SNAPSHOT_EVENT {
            self.depth.update_ask_depth(
                row.px,
                row.qty,
                row.local_ts
            );
        }
        // Processes a trade event
        else if row.ev & LOCAL_TRADE_EVENT == LOCAL_TRADE_EVENT {
            if self.trades.capacity() > 0 {
                self.trades.push(row.clone());
            }
        }

        // Checks
        let mut next_ts = 0;
        for rn in (self.row_num + 1)..self.data.len() {
            if self.data[rn].ev & LOCAL_EVENT == LOCAL_EVENT {
                self.row_num = rn;
                next_ts = self.data[rn].local_ts;
                break;
            }
        }

        if next_ts <= 0 {
            let next_data = unsafe {
                self.reader.next()?
            };
            let next_row = &next_data[0];
            next_ts = next_row.local_ts;
            let data = mem::replace(&mut self.data, next_data);
            self.reader.release(data);
            self.row_num = 0;
        }
        Ok((next_ts, i64::MAX))
    }

    fn process_recv_order(&mut self, timestamp: i64, wait_resp: i64) -> Result<i64, Error> {
        // Processes the order part.
        let mut next_timestamp = i64::MAX;
        while self.orders_from.len() > 0 {
            let recv_timestamp = self.orders_from.get_head_timestamp().unwrap();
            if timestamp == recv_timestamp {
                let order = self.orders_from.remove(0);
                next_timestamp = self.process_recv_order_(
                    order,
                    recv_timestamp,
                    wait_resp,
                    next_timestamp
                )?;
            } else {
                assert!(recv_timestamp > timestamp);
                break;
            }
        }
        Ok(next_timestamp)
    }

    fn frontmost_recv_order_timestamp(&self) -> i64 {
        self.orders_from.frontmost_timestamp()
    }

    fn frontmost_send_order_timestamp(&self) -> i64 {
        self.orders_to.frontmost_timestamp()
    }
}